File: backend/app/tts/earcons.py
"""Earcons - Idea11y Section 4.3.

Idea11y played a beep when a screen reader user arrived at a note a collaborator was working on,
choosing a non-speech cue "to minimize disruption in the user's workflow", and let users pick
earcon, speech, both, or none.

EchoNotes is single-user, so the same machinery signals capture and outline state instead. The
four-way feedback setting is kept exactly as the paper offered it.
"""

from __future__ import annotations

from enum import Enum


class Earcon(str, Enum):
    RECORDING_START = "recording_start"
    RECORDING_STOP = "recording_stop"
    NOTE_SAVED = "note_saved"
    NOTE_MOVED = "note_moved"
    LIST_END = "list_end"
    ERROR = "error"


class FeedbackMode(str, Enum):
    EARCON = "earcon"
    SPEECH = "speech"
    BOTH = "both"
    NONE = "none"


#: Spoken equivalent for every earcon, used in SPEECH and BOTH modes.
SPOKEN_EQUIVALENT: dict[Earcon, str] = {
    Earcon.RECORDING_START: "Recording",
    Earcon.RECORDING_STOP: "Stopped",
    Earcon.NOTE_SAVED: "Note saved",
    Earcon.NOTE_MOVED: "Note moved",
    Earcon.LIST_END: "End of list",
    Earcon.ERROR: "Error",
}


def feedback_for(earcon: Earcon, mode: FeedbackMode) -> dict[str, str | None]:
    """Return {"earcon": name or None, "speech": text or None} for the current mode."""
    raise NotImplementedError

File: frontend/src/a11y/earcons.ts
/**
 * Earcons - Idea11y Section 4.3.
 *
 * Idea11y used a beep instead of speech "to minimize disruption in the user's workflow", and let
 * users choose earcon, speech, both, or none. The same four-way setting applies here; only what
 * is being signalled differs, since EchoNotes has no collaborators.
 *
 * Tones are synthesised with the Web Audio API rather than loaded as files: no network wait, so
 * the recording-start cue lands the instant the key goes down.
 */

export type EarconName =
  | "recording_start"
  | "recording_stop"
  | "note_saved"
  | "note_moved"
  | "list_end"
  | "error";

export type FeedbackMode = "earcon" | "speech" | "both" | "none";

/** Frequency (Hz) and duration (ms) per cue. Rising = started, falling = finished. */
export const EARCON_TONES: Record<EarconName, { freq: number[]; ms: number }> = {
  recording_start: { freq: [660, 880], ms: 120 },
  recording_stop: { freq: [880, 660], ms: 120 },
  note_saved: { freq: [523, 784], ms: 140 },
  note_moved: { freq: [587, 587], ms: 100 },
  list_end: { freq: [440], ms: 90 },
  error: { freq: [311, 233], ms: 220 },
};

export function playEarcon(_name: EarconName, _mode: FeedbackMode): void {
  throw new Error("not implemented");
}
